### ISSUE TITLE
Segfault with game controller

### ISSUE THREAD

**Issue by Lisoph** [glfw/glfw#2320]
## Fix

[This line](https://github.com/glfw/glfw/blob/7482de6071d21db77a7236155da44c172a7f6c9e/src/cocoa_joystick.m#L186) is probably the culprit. Going by [this issue](https://github.com/panda3d/panda3d/issues/795) and their [fix](https://github.com/panda3d/panda3d/pull/797/files), `IOHIDDeviceCopyMatchingElements` can return null, which GLFW doesn't check for.

## I've noticed that GLFW segfaults when my controller is connected.

- GLFW v3.3.8
- macOS 13.3.1 (Apple Silicon)
- Xbox Elite Wireless Controller Series 2

To reproduce, compile the hello-world snippet from the website, then launch the app. You should be able to observe:

- When running without the controller connected, the app is running as expected.
- When running with the controller connected, the app immediately segfaults.

The expected behaviour is to not crash. GLFW should act like no controller is connected.

The issue seems to be related to macOS' input monitoring. To my understanding, when any program wants to read controller input, access has has to be explicitly granted by the user through macOS' settings. Until this access is granted for the hello-world program, GLFW segfaults.

Demo video of the issue:

https://user-images.githubusercontent.com/3426722/232581736-7eb08c13-0e41-46ad-b609-53d6189724b0.mov

In the video, access is granted to *Visual Studio Code* because the hello-world `./run` is just a child-process.

### Compiled with

```bash
clang++ -std=c++20 -g -Wall -Wextra -I/opt/homebrew/Cellar/glfw/3.3.8/include -L/opt/homebrew/Cellar/glfw/3.3.8/lib -o ./run *.cpp -lglfw
```

```bash
ᐅ  clang --version 
Apple clang version 14.0.3 (clang-1403.0.22.14.1)
Target: arm64-apple-darwin22.4.0
Thread model: posix
```

GLFW however was not compiled from source, but installed using Homebrew:

```bash
ᐅ  brew info glfw
==> glfw: stable 3.3.8 (bottled), HEAD
Multi-platform library for OpenGL applications
https://www.glfw.org/
/opt/homebrew/Cellar/glfw/3.3.8 (14 files, 490.4KB) *
  Poured from bottle using the formulae.brew.sh API on 2023-04-17 at 20:04:56
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/glfw.rb
```

### LLDB stack trace

```gdb
(lldb) thread backtrace
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x00000001952eaa28 CoreFoundation`CFArrayGetCount + 8
    frame #1: 0x0000000100097de4 libglfw.3.dylib`matchCallback + 504
    frame #2: 0x0000000198437d7c IOKit`__IOHIDManagerDeviceApplier + 592
    frame #3: 0x0000000195313bcc CoreFoundation`__CFSetApplyFunction_block_invoke + 28
    frame #4: 0x00000001953139e4 CoreFoundation`CFBasicHashApply + 148
    frame #5: 0x0000000195313928 CoreFoundation`CFSetApplyFunction + 320
    frame #6: 0x000000019843814c IOKit`__IOHIDManagerInitialEnumCallback + 120
    frame #7: 0x000000019534a710 CoreFoundation`__CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 28
    frame #8: 0x000000019534a6a4 CoreFoundation`__CFRunLoopDoSource0 + 176
    frame #9: 0x000000019534a414 CoreFoundation`__CFRunLoopDoSources0 + 244
    frame #10: 0x000000019534901c CoreFoundation`__CFRunLoopRun + 828
    frame #11: 0x000000019534858c CoreFoundation`CFRunLoopRunSpecific + 612
    frame #12: 0x0000000100097b68 libglfw.3.dylib`_glfwInitJoysticksNS + 472
    frame #13: 0x00000001000977dc libglfw.3.dylib`_glfwPlatformInit + 1204
    frame #14: 0x0000000100092294 libglfw.3.dylib`glfwInit + 80
    frame #15: 0x0000000100003ec4 run`main at main.cpp:4:10
    frame #16: 0x0000000194f13f28 dyld`start + 2236

(lldb) disassemble -frame
CoreFoundation`CFArrayGetCount:
    0x1952eaa20 <+0>:   tbz    x0, #0x3f, 0x1952eaa28    ; <+8>
    0x1952eaa24 <+4>:   b      0x1957a0c80               ; objc_msgSend$count
->  0x1952eaa28 <+8>:   ldr    x9, [x0]
    0x1952eaa2c <+12>:  adrp   x10, 369941
    0x1952eaa30 <+16>:  ldr    x10, [x10, #0x5a0]
    0x1952eaa34 <+20>:  bic    x8, x9, x10
    0x1952eaa38 <+24>:  ands   x16, x9, x10
    0x1952eaa3c <+28>:  b.eq   0x1952eaa64               ; <+68>
    0x1952eaa40 <+32>:  mov    x17, x0
    0x1952eaa44 <+36>:  movk   x17, #0x6ae1, lsl #48
    0x1952eaa48 <+40>:  autda  x16, x17
    0x1952eaa4c <+44>:  mov    x17, x16
    0x1952eaa50 <+48>:  xpacd  x17
    0x1952eaa54 <+52>:  cmp    x16, x17
    0x1952eaa58 <+56>:  b.eq   0x1952eaa60               ; <+64>
    0x1952eaa5c <+60>:  brk    #0xc472
    0x1952eaa60 <+64>:  orr    x8, x16, x8
    0x1952eaa64 <+68>:  cbz    x8, 0x1952eaab8           ; <+152>
    0x1952eaa68 <+72>:  adrp   x9, 375040
    0x1952eaa6c <+76>:  add    x9, x9, #0x158            ; __CFRuntimeBuiltinObjCClassTable + 152
    0x1952eaa70 <+80>:  ldr    x16, [x9]
    0x1952eaa74 <+84>:  mov    x17, x9
    0x1952eaa78 <+88>:  movk   x17, #0xcb85, lsl #48
    0x1952eaa7c <+92>:  autdb  x16, x17
    0x1952eaa80 <+96>:  mov    x17, x16
    0x1952eaa84 <+100>: xpacd  x17
    0x1952eaa88 <+104>: cmp    x16, x17
    0x1952eaa8c <+108>: b.eq   0x1952eaa94               ; <+116>
    0x1952eaa90 <+112>: brk    #0xc473
    0x1952eaa94 <+116>: cmp    x8, x16
    0x1952eaa98 <+120>: b.eq   0x1952eaab8               ; <+152>
    0x1952eaa9c <+124>: tbz    w8, #0x0, 0x1952eaa24     ; <+4>
    0x1952eaaa0 <+128>: adrp   x9, 369941
    0x1952eaaa4 <+132>: ldr    x9, [x9, #0x5a8]
    0x1952eaaa8 <+136>: ldr    x9, [x9]
    0x1952eaaac <+140>: and    x8, x9, x8
    0x1952eaab0 <+144>: cmp    x8, x16
    0x1952eaab4 <+148>: b.ne   0x1952eaa24               ; <+4>
    0x1952eaab8 <+152>: ldr    x0, [x0, #0x10]
    0x1952eaabc <+156>: ret    
(lldb) register read x0
      x0 = 0x0000000000000000
```

Looks like a classic null pointer dereference inside `CFArrayGetCount`. According to [StackOverflow](https://stackoverflow.com/a/54815725) `x0` is the register used for the first argument - in this case - and judging by [the docs](https://developer.apple.com/documentation/corefoundation/1388772-cfarraygetcount), this is most likely the pointer to the `CFArray`. So in short, seems like a NULL is passed to `CFArrayGetCount`.

### events.c output

None. The binary crashes before any output is written to stdout.

**Comment by Lisoph** [comment]
Fixed in commit 2c1d310

**Comment by Lisoph** [comment]
I can confirm that 2c1d310 fixes the segfault.

Thank you for merging my PR!
